exotations/solvers/exotica_ddp_solver/include/exotica_ddp_solver/abstract_ddp_solver.h

```diff
@@ -0,0 +1,219 @@
+//  * Neither the name of  nor the names of its contributors may be used to
+//    endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+
+#ifndef EXOTICA_DDP_SOLVER_ABSTRACT_DDP_SOLVER_H_
+#define EXOTICA_DDP_SOLVER_ABSTRACT_DDP_SOLVER_H_
+
+#include <exotica_core/motion_solver.h>
+#include <exotica_core/problems/dynamic_time_indexed_shooting_problem.h>
+#include <exotica_core/tools/conversions.h>
+
+namespace exotica
+{
+// \brief Base DDP Solver class that implements the forward pass.
+//  and utility functions. This is a templated class since we need
+//  the parameters from the solvers but do not have parameters here.
+//
+//  This is why this is a header-only
+template <typename Initializer>
+class AbstractDDPSolver : public MotionSolver
+{
+public:
+    ///\brief Solves the problem
+    ///@param solution Returned solution trajectory as a vector of joint configurations.
+    void Solve(Eigen::MatrixXd& solution) override;
+
+    ///\brief Binds the solver to a specific problem which must be pre-initalised
+    ///@param pointer Shared pointer to the motion planning problem
+    ///@return        Successful if the problem is a valid DynamicTimeIndexedProblem
+    void SpecifyProblem(PlanningProblemPtr pointer) override;
+
+protected:
+    DynamicTimeIndexedShootingProblemPtr prob_;       ///!< Shared pointer to the planning problem.
+    DynamicsSolverPtr dynamics_solver_;               ///!< Shared pointer to the dynamics solver.
+    std::vector<Eigen::MatrixXd> K_gains_, k_gains_;  ///!< Control gains.
+
+    ///\brief Computes the control gains for a the trajectory in the associated
+    ///     DynamicTimeIndexedProblem.
+    virtual void BackwardPass() = 0;
+
+    ///\brief Forward simulates the dynamics using the gains computed in the
+    ///     last BackwardPass;
+    /// @param alpha The learning rate.
+    /// @param ref_trajectory The reference state trajectory.
+    /// @return The cost associated with the new control and state trajectory.
+    double ForwardPass(double alpha, Eigen::MatrixXdRef ref_trajectory);
```

Comment:
`const double` - it may also be a `const` since it only modifies the problem, not the solver. `alpha` isn't the learning rate but the line-search parameter (but yeah, it's kind of the same thing).